cachito/web/models.py

```diff
@@ -930,6 +935,21 @@ def _state_to_json(state):
         rv.update(latest_state)
         return rv
 
+    def _sort_pkgs_in_place(self, packages):
+        """
+        Sorts lists of packages in place.
+
+        Sorting order: name -> version -> type.
+        If a package has a "dependencies" list, the packages inside it will be sorted as well.
+
+        :param list packages: the list of packages
+        """
+        packages.sort(key=lambda x: (x["name"], x["version"], x["type"]))
+
+        for package in packages:
+            if "dependencies" in package:
+                self._sort_pkgs_in_place(package["dependencies"])
```

Comment:
How about sort deps directly like this

```
sort_key = lambda x: (x["name"], x["version"], x["type"])
package["dependencies"].sort(sort_key)
```

Just thinking there seems no extra benefit to have one level recursion to sort the deps.

Dependency also has `dev`. Should it be added to the sort key for sorting deps?

Either packages or dependecies can't have two ones with same name, version and type, remove the version from sort key?

For the form of the sort key in a tuple, does the elements order matter for the sort? IMO, it would be good to sort by the type firstly, then the name. So, `(name, type)` or `(type, name)`, which one works for it?

Reply:
+1 for sorting directly, `dependencies` do not have their own `dependencies` so recursion is not needed
+1 for sorting also by `dev` for dependencies

I don't think we can remove `version` from the sorting key. Consider this:
```json
{
  "dependencies": [
    {"name": "requests", "version": "2.25.0", "type": "pip", "dev": false},
    {"name": "requests", "version": "2.24.0", "type": "pip", "dev": false}
  ]
}
```
^ this is a valid case, and we need to make sure the order will be deterministic.

I like the idea of sorting by type first. I'd say there is also an argument for sorting by name first, as that puts `go-package` packages closer to their corresponding `gomod` modules. But type would make more sense for requests that use multiple package managers. Ultimately we just need the order to be deterministic, I have no strong preference either way.

Reply:
So, I went with the `(type, dev, name, version)` order. I also got rid of the recursion :)
Let me know if you're ok it.